### ISSUE TITLE
Add swoop race layout randomization

### ISSUE THREAD

**Issue by glasnonck** [LaneDibello/Kotor-Randomizer-2#24]
Swoop randomization would involve changing the position of the booster pads and obstacles that are on the course. The number of boosters or obstacles may be variable, but further testing is needed. The max and min locations of the objects is not yet known, but a simple solution would be to restrict it to the max and min (x,y,z) of all objects in each layout file.

We could add a new randomization category for Layouts, but as of right now the primary utility of it would be to generate more interesting swoop races. Once layout randomization works for other modules (in any way) it can be separated into its own category.

Once the randomization is finalized, the help files need to be updated with documentation on the limits of the feature.

**Comment by glasnonck** [comment]
The layout file for each swoop race contains three lists: room objects, track objects, and obstacle objects. Room objects describe the environment or skybox that visually depicts the track. Track objects describe the on-rails portion of the track itself and the location of the boosters. Obstacle objects describe the location of the obstacle hitboxes but not the visual location of these obstacles. The main objects of interest are those described in the Track objects list. The room list will likely not be modified. The obstacle list may be modified in the future once we determine how to adjust the visual location of the obstacles as well.

The first of the track list controls the starting position of the swoop bike, and altering it changes where the bounds of the swoop track are within the level. For this reason, this item should not be modified until we understand it further.

The remaining objects within the track list describe the location of the booster pads on the track. Tatooine and Manaan have 30 boosters available (numbered 02 - 31), while Taris has 29 (number 09 is unused even if it is added to the layout list). All of these boosters can be rearranged throughout the track to provide a different experience. However, on both Tatooine and Manaan, booster 09 is placed in a bad location - well beyond the bounds of the track itself. Below is the boundary information I've gathered from looking at the layout files of each swoop race.

- Taris: min (81.48833, 190.006), max (122.339, 3742.73)
- Tatooine: min (80.796, 209.04), max (119.461, 3778.61)
- Manaan: min (84.7235, 170.042), max (123.842, 3589.13)
- Overall: min (80.796, 170.042), max (123.842, 3778.61)

I'll investigate to see if the overall min and max will be suitable for each track to increase variability. If it isn't suitable, the track specific min and max will be used as boundaries for random placement.

It may be possible to add the unused Taris booster by modifying some of the RIM files, but I haven't investigated this option. The ARE file doesn't include 09 in the list of minigame enemies with the rest of the boosters, so it could be as simple as adding it there.

I have also tried to increase the number of booster pads by simply adding more to the layout list with duplicate names. This didn't work, and only the first set of names were used to place boosters. Like I suggested above with the missing Taris booster, perhaps modifying the ARE file could allow us to place more - using boosters from other swoop tracks as well for a potential max of 90 boosters.

**Comment by glasnonck** [comment]
This issue isn't concerned with KotOR 2, but this seems to be the best place to report my testing on the topic.

Did some testing on swoop randomization in KotOR 2. The layout file is essentially the same, except the track list includes mines as well as boosters and the obstacle list contains different types of obstacles like walls. The model and hitbox of mines and boosters are moved with randomization, but the walls and other obstacles are not.

More testing is needed to integrate KotOR 2 into the randomizer, as I was unable to follow the same sequence when connecting to Layouts.bif to pull the raw game file. Instead, I just used the Layout class to open a file I extracted using Kotor Tool.
